adcc/AdcBlockView.py

```diff
@@ -0,0 +1,120 @@
+
+import libadcc
+
+
+class AdcBlockView(AdcMatrixlike):
+    def __init__(self, fullmatrix, block):
+        """
+        Object which acts as a single block of the ADC matrix.
+        """
+        if block not in fullmatrix.blocks:
+            raise ValueError(f"Block {block} not a valid block of passed "
+                             "matrix {self.__fullmatrix}")
+        self.__block = block
+        self.__fullmatrix = fullmatrix
+        super().__init__(fullmatrix)
+
+    def __repr__(self):
+        return f"AdcBlockView({self.__fullmatrix},{self.__block})"
+
+    @property
+    def blocks(self):
+        return [self.__block]
+
+    def has_block(self, block):
+        return self.__block == block
+
+    def __len__(self):
+        return np.prod([
+            self.mospaces.n_orbs(sp) for sp in self.block_spaces(self.__block)
+        ])
+
+    @property
+    def shape(self):
+        return (len(self), len(self))
+
+    def __assert_block(self, block):
+        if not all(b == self.__block for b in block):
+            raise ValueError(
+                f"The AdcBlockView onto the {self.__block} block of the matrix "
+                "{self.__fullmatrix} does not have block {block}."
+            )
+
+    def diagonal(self, block):
+        self.__assert_block(block)
+        return super().diagonal(block)
+
+    def block_spaces(self, block):
+        self.__assert_block(block)
+        return super().block_spaces(self.__block)
+
+    def compute_matvec(self, in_ampl, out_ampl=None):
+        """
+        Compute the matrix-vector product of the ADC matrix
+        with an excitation amplitude and return the result
+        in the out_ampl if it is given, else the result
+        will be returned.
+        """
+        # Unwrap input object, making sure we have two adcc.Tensors: one
+        # for the input and one for the result
+        return_bare_tensors = True
+        if isinstance(in_ampl, AmplitudeVector):
+            return_bare_tensors = False
+            if not in_ampl.blocks == [self.__block]:
+                raise ValueError("in_ampl does not consist of the correct blocks")
+            in_ampl = in_ampl[self.__block]
+        elif not isinstance(in_ampl, libadcc.Tensor):
+            raise TypeError("in_ampl has to be of type AmplitudeVector "
+                            "or Tensor.")
+        if out_ampl is None:
+            out_ampl = empty_like(in_ampl)
+        elif isinstance(out_ampl, AmplitudeVector):
+            if not out_ampl.blocks == [self.__block]:
+                raise ValueError("out_ampl does not consist of the "
+                                 "correct blocks")
+            out_ampl = out_ampl[self.__block]
+        elif not isinstance(out_ampl, libadcc.Tensor):
+            raise TypeError("out_ampl has to be of type AmplitudeVector "
+                            "or Tensor.")
+
+        # Compute the matrix-vector product
+        super().compute_apply(2 * self.__block, in_ampl, out_ampl)
```

Comment:
This looks as if one can only view a diagonal block (also fits the description of the PR)?
Then the class should be called `AdcDiagonalBlockView`.

Reply:
I think than I'd scratch the `View` to make it a more handy name. How about `AdcDiagonalBlock`?

Reply:
I think it's okay like that, I was just a little bit confused.